core/src/main/scala/kafka/controller/ZookeeperClient.scala

```diff
@@ -0,0 +1,249 @@
+          throw new ZookeeperClientTimeoutException
+        }
+        nanos = isConnectedOrExpiredCondition.awaitNanos(nanos)
+        state = zooKeeper.getState
+      }
+      state.isConnected
+    }
+  }
+
+  def registerZNodeChangeHandler(zNodeChangeHandler: ZNodeChangeHandler): ExistsResponse = {
+    registerZNodeChangeHandlers(Seq(zNodeChangeHandler)).head
+  }
+
+  def registerZNodeChangeHandlers(handlers: Seq[ZNodeChangeHandler]): Seq[ExistsResponse] = {
+    handlers.foreach(handler => zNodeChangeHandlers.put(handler.path, handler))
+    val asyncRequests = handlers.map(handler => ExistsRequest(handler.path, null))
+    handle(asyncRequests).asInstanceOf[Seq[ExistsResponse]]
+  }
+
+  def unregisterZNodeChangeHandler(path: String): Unit = {
+    zNodeChangeHandlers.remove(path)
+  }
+
+  def registerZNodeChildChangeHandler(zNodeChildChangeHandler: ZNodeChildChangeHandler): GetChildrenResponse = {
+    registerZNodeChildChangeHandlers(Seq(zNodeChildChangeHandler)).head
+  }
+
+  def registerZNodeChildChangeHandlers(handlers: Seq[ZNodeChildChangeHandler]): Seq[GetChildrenResponse] = {
+    handlers.foreach(handler => zNodeChildChangeHandlers.put(handler.path, handler))
+    val asyncRequests = handlers.map(handler => GetChildrenRequest(handler.path, null))
+    handle(asyncRequests).asInstanceOf[Seq[GetChildrenResponse]]
+  }
+
+  def unregisterZNodeChildChangeHandler(path: String): Unit = {
+    zNodeChildChangeHandlers.remove(path)
+  }
+
+  def close(): Unit = inWriteLock(initializationLock) {
+    zNodeChangeHandlers.clear()
+    zNodeChildChangeHandlers.clear()
+    zooKeeper.close()
+  }
+
+  private def initialize(): Unit = {
+    if (!zooKeeper.getState.isAlive) {
+      var now = time.milliseconds()
+      val threshold = now + connectionTimeoutMs
+      while (now < threshold) {
+        try {
+          zooKeeper = new ZooKeeper(connectString, sessionTimeoutMs, ZookeeperClientWatcher)
+          if (waitUntilConnectedOrExpired(threshold - now, TimeUnit.MILLISECONDS)) {
+            return
+          }
+        } catch {
+          case _: Exception =>
+            zooKeeper.close()
+            now = time.milliseconds()
+            if (now < threshold) {
+              time.sleep(1000)
+              now = time.milliseconds()
+            }
+        }
+      }
+      stateChangeHandler.onConnectionTimeout
+    }
+  }
+
+  private object ZookeeperClientWatcher extends Watcher {
+    override def process(event: WatchedEvent): Unit = {
+      if (event.getPath == null) {
+        inLock(isConnectedOrExpiredLock) {
+          isConnectedOrExpiredCondition.signalAll()
+        }
+        if (event.getState == KeeperState.AuthFailed) {
+          stateChangeHandler.onAuthFailure
+        } else if (event.getState == KeeperState.Expired) {
+          inWriteLock(initializationLock) {
+            stateChangeHandler.beforeInitializingSession
+            initialize()
+            stateChangeHandler.afterInitializingSession
```

Comment:
Suppose we only hold the write lock for initialize. Now you can permit a user to make calls concurrently as `stateChangeHandler.afterInitializingSession` is being run. I'm not sure if this is a problem.